src/core/meshcop/dataset_manager.cpp

```diff
@@ -769,15 +769,19 @@ void DatasetManager::DestroySecurelyStoredKeys(void) const
 {
     for (const SecurelyStoredTlv &entry : kSecurelyStoredTlvs)
     {
-        Crypto::Storage::DestroyKey(entry.GetKeyRef(mType));
+        uint32_t iid = otInstanceGetIdx(reinterpret_cast<otInstance *>(&InstanceLocator::GetInstance()));
+        Crypto::Storage::KeyRef keyRef =  entry.GetKeyRef(mType) + (iid * Crypto::Storage::KeyRefMaxOffset);
+        Crypto::Storage::DestroyKey(keyRef);
     }
 }
 
 void DatasetManager::MoveKeysToSecureStorage(Dataset &aDataset) const
 {
     for (const SecurelyStoredTlv &entry : kSecurelyStoredTlvs)
     {
-        SaveTlvInSecureStorageAndClearValue(aDataset, entry.mTlvType, entry.GetKeyRef(mType));
+        uint32_t iid = otInstanceGetIdx(reinterpret_cast<otInstance *>(&InstanceLocator::GetInstance()));
+        Crypto::Storage::KeyRef keyRef =  entry.GetKeyRef(mType) + (iid * Crypto::Storage::KeyRefMaxOffset);
+        SaveTlvInSecureStorageAndClearValue(aDataset, entry.mTlvType, keyRef);
```

Comment:
Thanks, @seb-fae, for this PR.

The approach seems to add an offset (determined based on the instance index) to `KeyRef` values whenever the `Crypto::Storage` functions are used. While this might work, I have some concerns about its design and potential drawbacks:

*   **Code repetition:** This offset logic would need to be replicated wherever `KeyRef`s are used, increasing the risk of errors and inconsistencies (easy to miss).
*   **Unnecessary complexity:** The code adds overhead and complexity even when `OPENTHREAD_CONFIG_MULTIPLE_INSTANCE_ENABLE` is disabled (which is the more common case).

I suggest a more centralized and maintainable solution:

*   **`Crypto::Storage` as a class:** Transform `Crypto::Storage` into a class with methods.
*   **Instance-specific storage:** Include a `Crypto::Storage` instance within the `Instance` class, providing each instance with its dedicated storage object.
*   **Internal KeyRef offset handling:** Implement the offset logic within the `Storage` class methods (only when `OPENTHREAD_CONFIG_MULTIPLE_INSTANCE_ENABLE` is enabled), eliminating the need for external offset management.

Thoughts?

Reply:
Hi Abtin,

You are right ! What you propose would indeed be the best and cleanest solution but require more modifications. What i provided was only a working patch that make dual instance work with two different security contexts. Without this patch using two different security context was not possible and lead to errors.

I am not a C++ expert but i will try to transform Crypto::storage as a class and integrate it as a member of Instance.

Should i need to create a new pull request when it will be done?

Best regards,
Sebastien